conformance-test-framework/src/main/java/org/jspecify/conformance/ExpectedFact.java

```diff
@@ -108,35 +124,91 @@ static class Reader {
 
     private static final Pattern EXPECTATION_COMMENT =
         Pattern.compile(
-            "\\s*// "
+            "\\s*// ("
+                + "(test:name:(?<testName>.*))"
+                + "|"
                 + ("(?<fact>"
                     + FACT_PATTERNS.stream().map(Pattern::pattern).collect(joining("|"))
                     + ")")
-                + "\\s*");
+                + ")\\s*");
+
+    private static final CharMatcher ASCII_DIGIT = CharMatcher.inRange('0', '9');
+
+    private final Map<Long, String> facts = new HashMap<>();
+    private final List<String> errors = new ArrayList<>();
+
+    private Path file;
+    private @Nullable String testName;
+    private long lineNumber;
 
     /** Reads expected facts from lines in a file. */
     ImmutableList<ExpectedFact> readExpectedFacts(Path file, List<String> lines) {
+      this.file = file;
       ImmutableList.Builder<ExpectedFact> expectedFacts = ImmutableList.builder();
-      Map<Long, String> facts = new HashMap<>();
       ListIterator<String> i = lines.listIterator();
       while (i.hasNext()) {
         String line = i.next();
-        long lineNumber = i.nextIndex();
+        lineNumber = i.nextIndex();
         Matcher matcher = EXPECTATION_COMMENT.matcher(line);
         if (matcher.matches()) {
+          setTestName(matcher.group("testName"));
           String fact = matcher.group("fact");
           if (fact != null) {
             facts.put(lineNumber, fact.trim());
           }
         } else {
+          if (testName != null) {
+            check(!facts.isEmpty(), "no expected facts for test named %s", testName);
+          }
           facts.forEach(
               (factLineNumber, factText) ->
-                  expectedFacts.add(new ExpectedFact(file, lineNumber, factText, factLineNumber)));
+                  expectedFacts.add(
+                      new ExpectedFact(file, lineNumber, testName, factText, factLineNumber)));
           facts.clear();
+          testName = null;
         }
       }
-      // TODO(netdpb): Report an error if facts is not empty.
-      return expectedFacts.build();
+      return checkUniqueTestNames(expectedFacts.build());
+    }
+
+    private void setTestName(@Nullable String testName) {
+      if (testName == null) {
+        return;
+      }
+      check(this.testName == null, "test name already set");
+      check(facts.isEmpty(), "test name must come before assertions for a line");
+      this.testName = testName.trim();
```

Comment:
Should this first check that the test name isn't empty?

Reply:
Good catch!